bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java

```diff
@@ -8939,6 +8965,80 @@ private int updateDb_045_To_046(final Connection conn, final SplashManager splas
       return newDbVersion;
    }
 
+   /**
+    * DB version 46 -> 47 ... MT version 22.?
+    *
+    * @param conn
+    * @param splashManager
+    * @return
+    * @throws SQLException
+    */
+   private int updateDb_046_To_047(final Connection conn, final SplashManager splashManager) throws SQLException {
+
+      final int newDbVersion = 47;
+
+      logDbUpdate_Start(newDbVersion);
+      updateMonitor(splashManager, newDbVersion);
+
+      try (Statement stmt = conn.createStatement()) {
+
+// SET_FORMATTING_OFF
+
+            SQL.RenameCol(stmt,     TABLE_TOUR_DATA,     RENAMED__TOUR_AVG_TEMPERATURE__FROM,        RENAMED__TOUR_AVG_TEMPERATURE__INTO);
+            SQL.RenameCol(stmt,     TABLE_TOUR_DATA,     RENAMED__TOUR_MAX_TEMPERATURE__FROM,        RENAMED__TOUR_MAX_TEMPERATURE__INTO);
+            SQL.RenameCol(stmt,     TABLE_TOUR_DATA,     RENAMED__TOUR_MIN_TEMPERATURE__FROM,        RENAMED__TOUR_MIN_TEMPERATURE__INTO);
+            SQL.RenameCol(stmt,     TABLE_TOUR_DATA,     RENAMED__TOUR_ISWEATHERDATAFROMAPI__FROM,   RENAMED__TOUR_ISWEATHERDATAFROMAPI__INTO);
+            SQL.RenameCol(stmt,     TABLE_TOUR_DATA,     RENAMED__TOUR_WEATHER_CLOUDS__FROM,         RENAMED__TOUR_WEATHER_CLOUDS__INTO);
+            SQL.RenameCol(stmt,     TABLE_TOUR_DATA,     RENAMED__TOUR_WEATHER_WIND_DIRECTION__FROM, RENAMED__TOUR_WEATHER_WIND_DIRECTION__INTO);
+            SQL.RenameCol(stmt,     TABLE_TOUR_DATA,     RENAMED__TOUR_WEATHER_WIND_SPEED__FROM,     RENAMED__TOUR_WEATHER_WIND_SPEED__INTO);
+
+            // Add new columns
+            SQL.AddColumn_Float(stmt, TABLE_TOUR_DATA,   "weather_Temperature_Average", DEFAULT_0);               //$NON-NLS-1$
+            SQL.AddColumn_Float(stmt, TABLE_TOUR_DATA,   "weather_Temperature_Max", DEFAULT_0);                   //$NON-NLS-1$
+            SQL.AddColumn_Float(stmt, TABLE_TOUR_DATA,   "weather_Temperature_Min", DEFAULT_0);                   //$NON-NLS-1$
+
+// SET_FORMATTING_ON
+      }
+
+      logDbUpdate_End(newDbVersion);
+
+      return newDbVersion;
+   }
+
+   private void updateDb_046_to_047_DataUpdate(final Connection conn) throws SQLException {
+
+      final long startTime = System.currentTimeMillis();
+
+      final int dbDataVersion = 47;
+
+      if (getDbVersion(conn, TABLE_DB_VERSION_DATA) >= dbDataVersion) {
+         // data version is higher -> nothing to do
+         return;
+      }
+
+      PreparedStatement stmtUpdate = null;
+
+      try {
+
+         stmtUpdate = conn.prepareStatement(UI.EMPTY_STRING
+
+               + "UPDATE " + TABLE_TOUR_DATA //                      //$NON-NLS-1$
```

Comment:
@wolfgang-ch 
This is the logic I set in place when the upgrade happens (46 ->47) to differentiate data from the device than data previously retrieved from the provider.
Thoughts ?

Reply:
For me it looks OK

Reply:
In the case where a tour has weather data manually entered/from provider but also has a `temperatureSerie`, after the `46 -> 47` database migration, the device average, max & min temperature values will have the same values as in the average/max & min from provider.
But the user can always recalculate the device values with this existing contextual menu option:
![image](https://user-images.githubusercontent.com/25235665/157502251-762ee996-66d1-4287-b492-674f5f0e0b8e.png)

Reply:
>In the case where a tour has weather data manually entered/from provider but also has a temperatureSerie, after the 46 -> 47 database migration, the device average, max & min temperature values will have the same values as in the average/max & min from provider.

This would be a bad behavior, after testing, the device avg/min/max are still the same as before.

The device avg/min/max values must be set only from the temperature serie values and in a few cases it is set from the device import

After manual edit

![image](https://user-images.githubusercontent.com/1283445/157813831-c7cbaf47-6062-41ca-b701-d7110d485238.png)

After loading weather data

![image](https://user-images.githubusercontent.com/1283445/157814640-b65d3578-3844-4475-9c4e-9a95a1f6ddd9.png)

Reply:
The only solution I see would be to trigger the computation of the device avg/min/max when performing the 46 -> 47 database upgrade but is that something possible during a database upgrade ??

Reply:
or maybe you meant they should be set to 0 and then recomputed manually by the user ?

Reply:
I do not really understand what should be set and when?

Maybe the user can decide in preferences when to set/overwrite values?

>The only solution I see would be to trigger the computation of the device avg/min/max when performing the 46 -> 47 database upgrade but is that something possible during a database upgrade ??

This `net.tourbook.database.TourDatabase.updateDb_042_To_043_DataUpdate_Concurrent()` is modifying tour data during a db update but concurrently, otherwise it is very slow

Reply:
> I do not really understand what should be set and when?
> 

You said this yesterday so i understood that you wanted me to recompute the temperature values from the device after the update ? or did I misunderstand ??
>This would be a bad behavior, after testing, the device avg/min/max are still the same as before.


> This `net.tourbook.database.TourDatabase.updateDb_042_To_043_DataUpdate_Concurrent()` is modifying tour data during a db update but concurrently, otherwise it is very slow

I will look at that when you confirm what needs to be done.

Reply:
I think there is some misunderstanding, this conversation started from your comment:

>In the case where a tour has weather data manually entered/from provider but also has a temperatureSerie, after the 46 -> 47 database migration, **the device average, max & min temperature values will have the same values as in the average/max & min from provider.**

Then I tested it but it was not as you had written (min/max/avg device data were still available), which was OK for me and there was no need to modify it.

Reply:
Maybe there is a misunderstanding, I'm going to try to explain better:

Here is the scenario of what will happen when going from DB 46 -> 47 with the current database code SQL query below):

1) If the tour has the boolean `isWeatherDataFromProvider` set to **true**, then the **current** (DB 46) values are copied into the new fields weather_Temperature_Average, Max & Min FROM the older weather_Temperature_Average_Device, Max, Min (Remember, in DB 46, there was only 1 temperature field that could have the value from the temperatureserie or manually entered from provider so we need some logic to know which value to put in which field)

2) If the tour has the boolean `isWeatherDataFromProvider` set to **false**, then we keep the values in weather_Temperature_Average_Device as we assume those values are from the device, so there is nothing to do

- If scenario 1) happens, then the values in weather_Temperature_Average & weather_Temperature_Average_Device are the same after the update but don't necessarily represent the actual value from the device. But like I mentioned [above](https://github.com/mytourbook/mytourbook/pull/483#discussion_r822930455), the user can recompute the device value manually or I can inspire myself from your function `updateDb_042_To_043_DataUpdate_Concurrent` to recompute the temperature values from the device with the TourData function `computeAvg_Temperature()`

Does it make sense ? If yes, what do you think ?


```
"UPDATE TABLE_TOUR_DATA 
SET                                  
 weather_Temperature_Average = weather_Temperature_Average_Device,
 weather_Temperature_Max = weather_Temperature_Max_Device,
 weather_Temperature_Min = weather_Temperature_Min_Device
 WHERE isWeatherDataFromProvider=true"
```

Reply:
I overlooked the `WHERE isWeatherDataFromProvider=true"` and was also wondering, that my device values were not overwritten, sorry about that.

For scenario 1) you could either implement the concurrent tour data update solution (with the already existing showcase it should not too much work) or you should update the readme in the **Changes** section 

With your detailed description, now I could "fix" my thoughts :-)

Reply:
Ah ok, that makes sense now.

I have implemented the concurrent tour data update solution as I was able to inspire myself from `updateDb_042_To_043_DataUpdate_Concurrent`. Like you said, it was not too much work

Reply:
Fixed in my [PR](https://github.com/mytourbook/mytourbook/pull/485)